timApp/plugin/pluginControl.py

```diff
@@ -756,6 +756,10 @@ def pluginify(
             par = html_pars[idx]
             for plugin_key, plugin_html in par.plugin_htmls.items():
                 h = h.replace(plugin_key, plugin_html)
+            if h.startswith("<p>") and h.endswith("</p>"):
+                # dumbo wraps inline block with <p>-element, but the plugins may be lazy and have <div>-elements,
+                # breaking the <p> and preventing pluginloader from loading the lazy plugins correctly
+                h = h[3:-4]
```

Comment:
En ole ihan varma, onko tuo oikea ratkaisu, koska lohkossa voi olla useita kappaleita. Esimerkiksi

```` 
#- {defaultplugin="csPlugin"}
Jotain tekstiä
{#xinline1 type: text/tiny, lazy: true #}
{#xinline2 type: text/tiny , lazy: true #}

Jotain muuta tekstiä
{#xinlinelazy3 type: text/tiny, lazy: true  #}
{#xinlinelazy4 type: text/tiny, lazy: true  #}
{#xinlinelazy5 type: text/tiny, lazy: true  #}
{#xinlinelazy6:textfield lazy: true  #}
````

tekee kaksi kappaletta eli

```html
<p>
  Jotain tekstiä
  <!-- pluginit 1-2 -->
</p>
<p>
  Jotain muuta tekstiä
  <!-- pluginit 3-6 -->
</p>
```

Tällä logiikalla saa siis HTMLn epävalidiksi, jolloin eri selaimet käsittelevät eri tavoin. Ainakin Firefoxissa se silloin laittaa `<p>`:n ja sen sisään xinlinelazy3:n. Esimerkki:

https://timdevs01-3.it.jyu.fi/view/lazyinline2

Perimmäinen ongelma tässä tosiaankin on, että csPluginin lazy palauttaa `div`eja, vaikka `p`:ssä ei niitä saa olla (`p` saa sisältää vain tekstitason elementteja, `div` on rakenteellinen elementti). Itselleni tulee nyt mieleen vain kaksi tapaa edetä

1. Korjataan pluginien lazyjen `div`it esim. `span`ksi. Tämä taitaa vaatia jonkin verran säätöä, koska monessa pluginissa on juuri div. Tuo 
2. Korvataan inline-lohkossa kaikiki `p`t `div`:lla. Tämä vaatisi vähiten muutosta, mutta saattaakohan se sitten rikkoa jotain muuta.

Reply:
> En ole ihan varma, onko tuo oikea ratkaisu, koska lohkossa voi olla useita kappaleita. Esimerkiksi
> 
> ```
> #- {defaultplugin="csPlugin"}
> Jotain tekstiä
> {#xinline1 type: text/tiny, lazy: true #}
> {#xinline2 type: text/tiny , lazy: true #}
> 
> Jotain muuta tekstiä
> {#xinlinelazy3 type: text/tiny, lazy: true  #}
> {#xinlinelazy4 type: text/tiny, lazy: true  #}
> {#xinlinelazy5 type: text/tiny, lazy: true  #}
> {#xinlinelazy6:textfield lazy: true  #}
> ```
> 
> tekee kaksi kappaletta eli
> 
> ```
> <p>
>   Jotain tekstiä
>   <!-- pluginit 1-2 -->
> </p>
> <p>
>   Jotain muuta tekstiä
>   <!-- pluginit 3-6 -->
> </p>
> ```
> 
> Tällä logiikalla saa siis HTMLn epävalidiksi, jolloin eri selaimet käsittelevät eri tavoin. Ainakin Firefoxissa se silloin laittaa `<p>`:n ja sen sisään xinlinelazy3:n. Esimerkki:
> 
> https://timdevs01-3.it.jyu.fi/view/lazyinline2
> 
> Perimmäinen ongelma tässä tosiaankin on, että csPluginin lazy palauttaa `div`eja, vaikka `p`:ssä ei niitä saa olla (`p` saa sisältää vain tekstitason elementteja, `div` on rakenteellinen elementti). Itselleni tulee nyt mieleen vain kaksi tapaa edetä
> 
>     1. Korjataan pluginien lazyjen `div`it esim. `span`ksi. Tämä taitaa vaatia jonkin verran säätöä, koska monessa pluginissa on juuri div. Tuo
> 
>     2. Korvataan inline-lohkossa kaikiki `p`t `div`:lla. Tämä vaatisi vähiten muutosta, mutta saattaakohan se sitten rikkoa jotain muuta.

Joo, en huomannutkaan kokeilla monimutkaisempaa kappaletta. Laitan vielä toistaiseksi draftiksi.
Jos kokeilee noita `<p>`-elementtien korvaamista `<div>`eillä niin voiko tuossa dumbo-kutsussa suoraan sanoa dumbolle että ei kääri noita oletuksena `<p>` vaan käyttää `<div>`? Vai joutuuko nuo korvaamaan timin puolella dumbo-kutsun jälkeen